testnet/stacks-node/src/run_loop/nakamoto.rs

```diff
@@ -0,0 +1,1023 @@
+        );
+        self.set_globals(globals.clone());
+
+        // have headers; boot up the chains coordinator and instantiate the chain state
+        let coordinator_thread_handle = self.spawn_chains_coordinator(
+            &burnchain_config,
+            coordinator_receivers,
+            globals.get_miner_status(),
+        );
+        self.start_prometheus();
+
+        // We announce a new burn block so that the chains coordinator
+        // can resume prior work and handle eventual unprocessed sortitions
+        // stored during a previous session.
+        globals.coord().announce_new_burn_block();
+
+        // Make sure at least one sortition has happened, and make sure it's globally available
+        let sortdb = burnchain.sortdb_mut();
+        let (rc_aligned_height, sn) =
+            RunLoop::get_reward_cycle_sortition_db_height(&sortdb, &burnchain_config);
+
+        let burnchain_tip_snapshot = if sn.block_height == burnchain_config.first_block_height {
+            // need at least one sortition to happen.
+            burnchain
+                .wait_for_sortitions(globals.coord().clone(), sn.block_height + 1)
+                .expect("Unable to get burnchain tip")
+                .block_snapshot
+        } else {
+            sn
+        };
+
+        globals.set_last_sortition(burnchain_tip_snapshot.clone());
+
+        // Boot up the p2p network and relayer, and figure out how many sortitions we have so far
+        // (it could be non-zero if the node is resuming from chainstate)
+        let mut node = StacksNode::spawn(self, globals.clone(), relay_recv);
+        let liveness_thread = self.spawn_chain_liveness_thread(globals.clone());
+
+        // Wait for all pending sortitions to process
+        let burnchain_db = burnchain_config
+            .open_burnchain_db(false)
+            .expect("FATAL: failed to open burnchain DB");
+        let burnchain_db_tip = burnchain_db
+            .get_canonical_chain_tip()
+            .expect("FATAL: failed to query burnchain DB");
+        let mut burnchain_tip = burnchain
+            .wait_for_sortitions(globals.coord().clone(), burnchain_db_tip.block_height)
+            .expect("Unable to get burnchain tip");
+
+        // Start the runloop
+        debug!("Runloop: Begin run loop");
+        self.counters.bump_blocks_processed();
+
+        let mut sortition_db_height = rc_aligned_height;
+        let mut burnchain_height = sortition_db_height;
+        let mut num_sortitions_in_last_cycle;
+
+        // prepare to fetch the first reward cycle!
+        let mut target_burnchain_block_height = cmp::min(
+            burnchain_config.reward_cycle_to_block_height(
+                burnchain_config
+                    .block_height_to_reward_cycle(burnchain_height)
+                    .expect("BUG: block height is not in a reward cycle")
+                    + 1,
+            ),
+            burnchain.get_headers_height() - 1,
```

Comment:
I assume this can never underflow? I generally prefer to use saturating_sub just to be absolutely certain.